File: ios/Sources/WifiAwarePlugin/WifiAwarePlugin.swift
import Foundation
import Capacitor
import Network

@objc(WifiAwarePlugin)
public class WifiAwarePlugin: CAPPlugin, CAPBridgedPlugin, WifiAwareDelegate {
  public let identifier = "WifiAwarePlugin"
  public let jsName = "WifiAware"
  public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name: "attach", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "publish", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "stopPublish", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "subscribe", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "stopSubscribe", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "sendMessage", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "startSocket", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "stopSocket", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "sendFileTransfer", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "respondToFileTransfer", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "removeAllListeners", returnType: CAPPluginReturnPromise)
  ]

  // Set this to true once you've added the entitlement and declared services
  let ENABLE_WIFI_AWARE_BETA = false
  
  private var wifiAware: WifiAware?
  
  override public func load() {
    // Initialize the WiFi Aware implementation if enabled
    if ENABLE_WIFI_AWARE_BETA {
      wifiAware = WifiAware(delegate: self)
    }
  }
  
  @objc func attach(_ call: CAPPluginCall) {
    #if canImport(Network)
    if ENABLE_WIFI_AWARE_BETA, let wifiAware = wifiAware {
      let (available, reason) = wifiAware.checkAvailability()
      notifyStateChanged(available: available, reason: reason)
      call.resolve([
        "available": available,
        "reason": reason as Any
      ])
      return
    }
    #endif
    
    let reason = "iOS Wi-Fi Aware requires iOS 18+, entitlement, and declared services"
    notifyStateChanged(available: false, reason: reason)
    call.resolve([
      "available": false,
      "reason": reason
    ])
  }

  @objc func publish(_ call: CAPPluginCall) {
    guard ENABLE_WIFI_AWARE_BETA, let wifiAware = wifiAware else {
      call.reject("Wi-Fi Aware not available on iOS")
      return
    }
    
    guard let serviceName = call.getString("serviceName") else {
      call.reject("serviceName is required")
      return
    }
    
    // Extract optional parameters
    var serviceInfo: Data? = nil
    if let serviceInfoBase64 = call.getString("serviceInfoBase64") {
      serviceInfo = Data(base64Encoded: serviceInfoBase64)
    }
    
    let instantMode = call.getBool("instantMode") ?? false
    
    // Start publishing
    wifiAware.startPublishing(serviceName: serviceName, serviceInfo: serviceInfo, instantMode: instantMode)
    call.resolve()
  }

  @objc func stopPublish(_ call: CAPPluginCall) { 
    wifiAware?.stopPublishing()
    call.resolve() 
  }
  
  @objc func subscribe(_ call: CAPPluginCall) {
    guard ENABLE_WIFI_AWARE_BETA, let wifiAware = wifiAware else {
      call.reject("Wi-Fi Aware not available on iOS")
      return
    }
    
    guard let serviceName = call.getString("serviceName") else {
      call.reject("serviceName is required")
      return
    }
    
    let instantMode = call.getBool("instantMode") ?? false
    
    // Start browsing
    wifiAware.startBrowsing(serviceName: serviceName, instantMode: instantMode)
    call.resolve()
  }
  
  @objc func stopSubscribe(_ call: CAPPluginCall) { 
    wifiAware?.stopBrowsing()
    call.resolve() 
  }
  
  @objc func sendMessage(_ call: CAPPluginCall) {
    guard ENABLE_WIFI_AWARE_BETA, let wifiAware = wifiAware else {
      call.reject("Wi-Fi Aware not available on iOS")
      return
    }
    
    guard let peerId = call.getString("peerId"), 
          let dataBase64 = call.getString("dataBase64"),
          let data = Data(base64Encoded: dataBase64) else {
      call.reject("peerId and dataBase64 are required")
      return
    }
    
    wifiAware.sendMessage(peerId: peerId, data: data)
    call.resolve()
  }
  
  @objc func startSocket(_ call: CAPPluginCall) {
    guard ENABLE_WIFI_AWARE_BETA, let wifiAware = wifiAware else {
      call.reject("Wi-Fi Aware not available on iOS")
      return
    }
    
    guard let peerId = call.getString("peerId"), 
          let pskPassphrase = call.getString("pskPassphrase") else {
      call.reject("peerId and pskPassphrase are required")
      return
    }
    
    let asServer = call.getBool("asServer") ?? false
    
    wifiAware.startSocket(peerId: peerId, pskPassphrase: pskPassphrase, asServer: asServer)
    call.resolve()
  }
  
  @objc func stopSocket(_ call: CAPPluginCall) {
    if let peerId = call.getString("peerId") {
      wifiAware?.stopSocket(peerId: peerId)
    } else {
      wifiAware?.stopAllSockets()
    }
    call.resolve()
  }
  
  @objc func sendFileTransfer(_ call: CAPPluginCall) {
    guard ENABLE_WIFI_AWARE_BETA, let wifiAware = wifiAware else {
      call.reject("Wi-Fi Aware not available on iOS")
      return
    }
    
    guard let peerId = call.getString("peerId") else {
      call.reject("peerId is required")
      return
    }
    
    let filePath = call.getString("filePath")
    let fileBase64 = call.getString("fileBase64")
    
    if (filePath == nil && fileBase64 == nil) || (filePath != nil && fileBase64 != nil) {
      call.reject("Either filePath OR fileBase64 must be provided")
      return
    }
    
    var fileName = call.getString("fileName")
    
    if fileName == nil && filePath != nil {
      // Extract filename from path
      let url = URL(fileURLWithPath: filePath!)
      fileName = url.lastPathComponent
    }
    
    if fileName == nil {
      call.reject("fileName is required when using fileBase64")
      return
    }
    
    let mimeType = call.getString("mimeType")
    
    do {
      let transferId = try wifiAware.sendFileTransfer(
        peerId: peerId,
        filePath: filePath,
        fileBase64: fileBase64,
        fileName: fileName!,
        mimeType: mimeType
      )
      
      call.resolve(["transferId": transferId])
    } catch {
      call.reject("Failed to initiate file transfer: \(error.localizedDescription)")
    }
  }
  
  @objc func respondToFileTransfer(_ call: CAPPluginCall) {
    guard ENABLE_WIFI_AWARE_BETA, let wifiAware = wifiAware else {
      call.reject("Wi-Fi Aware not available on iOS")
      return
    }
    
    guard let peerId = call.getString("peerId"),
          let transferId = call.getString("transferId") else {
      call.reject("peerId and transferId are required")
      return
    }
    
    let accept = call.getBool("accept") ?? false
    let savePath = call.getString("savePath")
    
    wifiAware.respondToFileTransfer(peerId: peerId, transferId: transferId, accept: accept, savePath: savePath)
    call.resolve()
  }
  
  @objc override func removeAllListeners(_ call: CAPPluginCall) {
    super.removeAllListeners(call)
    call.resolve()
  }
  
  // MARK: - WifiAwareDelegate Methods
  
  public func stateChanged(available: Bool, reason: String?) {
    notifyStateChanged(available: available, reason: reason)
  }
  
  public func serviceFound(peerId: String, serviceName: String, serviceInfoBase64: String?, distanceMm: Int?) {
    var data: [String: Any] = [
      "peerId": peerId,
      "serviceName": serviceName
    ]
    
    if let serviceInfoBase64 = serviceInfoBase64 {
      data["serviceInfoBase64"] = serviceInfoBase64
    }
    
    if let distanceMm = distanceMm {
      data["distanceMm"] = distanceMm
    }
    
    notifyListeners("serviceFound", data: data)
  }
  
  public func serviceLost(peerId: String, serviceName: String) {
    let data: [String: Any] = [
      "peerId": peerId,
      "serviceName": serviceName
    ]
    
    notifyListeners("serviceLost", data: data)
  }
  
  public func messageReceived(peerId: String, dataBase64: String) {
    let data: [String: Any] = [
      "peerId": peerId,
      "dataBase64": dataBase64
    ]
    
    notifyListeners("messageReceived", data: data)
  }
  
  public func socketReady(role: String, localPort: Int?, peerIpv6: String?, peerPort: Int?) {
    var data: [String: Any] = ["role": role]
    
    if let localPort = localPort {
      data["localPort"] = localPort
    }
    
    if let peerIpv6 = peerIpv6 {
      data["peerIpv6"] = peerIpv6
    }
    
    if let peerPort = peerPort {
      data["peerPort"] = peerPort
    }
    
    notifyListeners("socketReady", data: data)
  }
  
  public func socketClosed(socketId: String?) {
    var data: [String: Any] = [:]
    if let socketId = socketId {
      data["socketId"] = socketId
    }
    notifyListeners("socketClosed", data: data)
  }
  
  public func fileTransferRequest(peerId: String, transferId: String, fileName: String, mimeType: String?, fileSize: Int64) {
    var data: [String: Any] = [
      "peerId": peerId,
      "transferId": transferId,
      "fileName": fileName,
      "fileSize": fileSize
    ]
    
    if let mimeType = mimeType {
      data["mimeType"] = mimeType
    }
    
    notifyListeners("fileTransferRequest", data: data)
  }
  
  public func fileTransferProgress(peerId: String, transferId: String, fileName: String, bytesTransferred: Int64, totalBytes: Int64, direction: String, status: String) {
    let progress = totalBytes > 0 ? Int(bytesTransferred * 100 / totalBytes) : 0
    
    let data: [String: Any] = [
      "peerId": peerId,
      "transferId": transferId,
      "fileName": fileName,
      "bytesTransferred": bytesTransferred,
      "totalBytes": totalBytes,
      "progress": progress,
      "direction": direction,
      "status": status
    ]
    
    notifyListeners("fileTransferProgress", data: data)
  }
  
  public func fileTransferCompleted(peerId: String, transferId: String, fileName: String, filePath: String?, fileBase64: String?) {
    var data: [String: Any] = [
      "peerId": peerId,
      "transferId": transferId,
      "fileName": fileName
    ]
    
    if let filePath = filePath {
      data["filePath"] = filePath
    }
    
    if let fileBase64 = fileBase64 {
      data["fileBase64"] = fileBase64
    }
    
    notifyListeners("fileTransferCompleted", data: data)
  }
  
  public func peerConnected(socketId: String, peerId: String, deviceInfo: [String: Any]?) {
    var data: [String: Any] = [
      "socketId": socketId,
      "peerId": peerId
    ]
    
    if let deviceInfo = deviceInfo {
      data["deviceInfo"] = deviceInfo
    }
    
    notifyListeners("peerConnected", data: data)
  }
  
  public func peerDisconnected(socketId: String, peerId: String) {
    let data: [String: Any] = [
      "socketId": socketId,
      "peerId": peerId
    ]
    
    notifyListeners("peerDisconnected", data: data)
  }
  
  // MARK: - Helper Methods
  
  private func notifyStateChanged(available: Bool, reason: String?) {
    var data: [String: Any] = ["available": available]
    
    if let reason = reason {
      data["reason"] = reason
    }
    
    notifyListeners("stateChanged", data: data)
  }
}

File: android/src/main/java/com/asaf/plugins/wifiaware/WifiAwarePlugin.java
package com.asaf.plugins.wifiaware;

import android.Manifest;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.os.Build;

import androidx.core.app.ActivityCompat;

import com.getcapacitor.BridgeActivity;
import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

@CapacitorPlugin(name = "WifiAware", permissions = {
        @Permission(strings = { Manifest.permission.NEARBY_WIFI_DEVICES }, alias = "nearby"),
        @Permission(strings = { Manifest.permission.ACCESS_FINE_LOCATION }, alias = "location")
})
public class WifiAwarePlugin extends Plugin implements WifiAwareShim.MessageSink {

    private WifiAwareShim aware;
    private WifiAwareStateReceiver stateReceiver;

    @Override
    public void load() {
        aware = new WifiAwareShim((BridgeActivity) getActivity(), this);

        stateReceiver = new WifiAwareStateReceiver(result -> {
            notifyListeners("stateChanged", resultToJS(result));
        });

        IntentFilter filter = new IntentFilter("android.net.wifi.aware.action.WIFI_AWARE_STATE_CHANGED");
        getContext().registerReceiver(stateReceiver, filter);
    }

    @Override
    protected void handleOnDestroy() {
        try {
            getContext().unregisterReceiver(stateReceiver);
        } catch (Throwable ignore) {
        }
        if (aware != null)
            aware.destroy();
    }

    // ==== API ====

    @PluginMethod
    public void attach(PluginCall call) {
        WifiAwareStateReceiver.AttachResult res = aware.attach();
        notifyListeners("stateChanged", resultToJS(res));
        call.resolve(resultToJS(res));
    }

    @PluginMethod
    public void publish(PluginCall call) {
        final PublishOptions opts;
        try {
            opts = PublishOptions.fromCall(call);
        } catch (IllegalArgumentException e) {
            call.reject(e.getMessage());
            return;
        }
        ensurePermissions(call, () -> {
            try {
                aware.publish(opts, peer -> notifyListeners("serviceFound", peerToJS(peer, opts.serviceName)));
                call.resolve();
            } catch (Exception e) {
                call.reject(e.getMessage());
            }
        });
    }

    @PluginMethod
    public void stopPublish(PluginCall call) {
        aware.stopPublish();
        call.resolve();
    }

    @PluginMethod
    public void subscribe(PluginCall call) {
        final SubscribeOptions opts;
        try {
            opts = SubscribeOptions.fromCall(call);
        } catch (IllegalArgumentException e) {
            call.reject(e.getMessage());
            return;
        }
        ensurePermissions(call, () -> {
            try {
                aware.subscribe(
                        opts,
                        peer -> notifyListeners("serviceFound", peerToJS(peer, opts.serviceName)),
                        peerId -> {
                            JSObject js = new JSObject();
                            js.put("peerId", peerId);
                            js.put("serviceName", opts.serviceName);
                            notifyListeners("serviceLost", js);
                        });
                call.resolve();
            } catch (Exception e) {
                call.reject(e.getMessage());
            }
        });
    }

    @PluginMethod
    public void stopSubscribe(PluginCall call) {
        aware.stopSubscribe();
        call.resolve();
    }

    @PluginMethod
    public void sendMessage(PluginCall call) {
        String peerId = call.getString("peerId");
        String dataBase64 = call.getString("dataBase64");
        if (peerId == null || dataBase64 == null) {
            call.reject("peerId and dataBase64 required");
            return;
        }
        try {
            aware.sendMessage(peerId, dataBase64);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void startSocket(PluginCall call) {
        String peerId = call.getString("peerId");
        String psk = call.getString("pskPassphrase");
        boolean asServer = call.getBoolean("asServer", false);
        if (peerId == null || psk == null) {
            call.reject("peerId and pskPassphrase required");
            return;
        }
        try {
            aware.startSocket(peerId, psk, asServer,
                    info -> {
                        notifyListeners("socketReady", socketToJS(info));
                        call.resolve(socketToJS(info));
                    },
                    () -> notifyListeners("socketClosed", new JSObject()));
        } catch (Exception e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void stopSocket(PluginCall call) {
        aware.stopSocket();
        notifyListeners("socketClosed", new JSObject());
        call.resolve();
    }
    
    @PluginMethod
    public void sendFileTransfer(PluginCall call) {
        String peerId = call.getString("peerId");
        String filePath = call.getString("filePath");
        String fileBase64 = call.getString("fileBase64");
        String fileName = call.getString("fileName");
        String mimeType = call.getString("mimeType");
        
        if (peerId == null) {
            call.reject("peerId is required");
            return;
        }
        
        if ((filePath == null && fileBase64 == null) || (filePath != null && fileBase64 != null)) {
            call.reject("Either filePath OR fileBase64 must be provided");
            return;
        }
        
        if (fileName == null) {
            if (filePath != null) {
                // Extract filename from path
                int lastPathSeparator = filePath.lastIndexOf('/');
                if (lastPathSeparator < 0) {
                    lastPathSeparator = filePath.lastIndexOf('\\');
                }
                if (lastPathSeparator >= 0 && lastPathSeparator < filePath.length() - 1) {
                    fileName = filePath.substring(lastPathSeparator + 1);
                } else {
                    call.reject("fileName is required when the filename cannot be determined from filePath");
                    return;
                }
            } else {
                call.reject("fileName is required when using fileBase64");
                return;
            }
        }
        
        try {
            String transferId = aware.sendFileTransfer(peerId, filePath, fileBase64, fileName, mimeType);
            JSObject result = new JSObject();
            result.put("transferId", transferId);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to initiate file transfer: " + e.getMessage());
        }
    }
    
    @PluginMethod
    public void respondToFileTransfer(PluginCall call) {
        String peerId = call.getString("peerId");
        String transferId = call.getString("transferId");
        Boolean accept = call.getBoolean("accept", false);
        String savePath = call.getString("savePath");
        
        if (peerId == null || transferId == null) {
            call.reject("peerId and transferId are required");
            return;
        }
        
        try {
            aware.respondToFileTransfer(peerId, transferId, accept, savePath);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to respond to file transfer: " + e.getMessage());
        }
    }

    @PluginMethod
    public void removeAllListeners(PluginCall call) {
        super.removeAllListeners(call);
        call.resolve();
    }

    // ==== MessageSink from shim ====

    @Override
    public void onMessageReceived(String peerId, String dataBase64) {
        JSObject js = new JSObject();
        js.put("peerId", peerId);
        js.put("dataBase64", dataBase64);
        notifyListeners("messageReceived", js);
    }
    
    @Override
    public void onFileTransferRequest(String peerId, String transferId, String fileName, String mimeType, long fileSize) {
        JSObject js = new JSObject();
        js.put("peerId", peerId);
        js.put("transferId", transferId);
        js.put("fileName", fileName);
        if (mimeType != null && !mimeType.isEmpty()) {
            js.put("mimeType", mimeType);
        }
        js.put("fileSize", fileSize);
        notifyListeners("fileTransferRequest", js);
    }
    
    @Override
    public void onFileTransferProgress(String peerId, String transferId, String fileName, 
                                       long bytesTransferred, long totalBytes, String direction, String status) {
        JSObject js = new JSObject();
        js.put("peerId", peerId);
        js.put("transferId", transferId);
        js.put("fileName", fileName);
        js.put("bytesTransferred", bytesTransferred);
        js.put("totalBytes", totalBytes);
        js.put("progress", (int)((bytesTransferred * 100) / Math.max(1, totalBytes)));
        js.put("direction", direction);
        js.put("status", status);
        notifyListeners("fileTransferProgress", js);
    }
    
    @Override
    public void onFileTransferCompleted(String peerId, String transferId, String fileName, 
                                       String filePath, String fileBase64) {
        JSObject js = new JSObject();
        js.put("peerId", peerId);
        js.put("transferId", transferId);
        js.put("fileName", fileName);
        if (filePath != null) {
            js.put("filePath", filePath);
        }
        if (fileBase64 != null) {
            js.put("fileBase64", fileBase64);
        }
        notifyListeners("fileTransferCompleted", js);
    }
    
    @Override
    public void onPeerConnected(String socketId, String peerId, Map<String, Object> deviceInfo) {
        JSObject js = new JSObject();
        js.put("socketId", socketId);
        js.put("peerId", peerId);
        
        if (deviceInfo != null) {
            JSObject deviceInfoJs = new JSObject();
            for (Map.Entry<String, Object> entry : deviceInfo.entrySet()) {
                deviceInfoJs.put(entry.getKey(), entry.getValue());
            }
            js.put("deviceInfo", deviceInfoJs);
        }
        
        notifyListeners("peerConnected", js);
    }
    
    @Override
    public void onPeerDisconnected(String socketId, String peerId) {
        JSObject js = new JSObject();
        js.put("socketId", socketId);
        js.put("peerId", peerId);
        notifyListeners("peerDisconnected", js);
    }
    
    @Override
    public void onSocketClosed(String socketId) {
        JSObject js = new JSObject();
        if (socketId != null) {
            js.put("socketId", socketId);
        }
        notifyListeners("socketClosed", js);
    }

    // ==== Permission handling ====

    private void ensurePermissions(PluginCall call, Runnable proceed) {
        if (Build.VERSION.SDK_INT >= 33) {
            if (ActivityCompat.checkSelfPermission(getContext(),
                    Manifest.permission.NEARBY_WIFI_DEVICES) != PackageManager.PERMISSION_GRANTED) {
                requestPermissionForAlias("nearby", call, "permCallback");
                return;
            }
        } else {
            if (ActivityCompat.checkSelfPermission(getContext(),
                    Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                requestPermissionForAlias("location", call, "permCallback");
                return;
            }
        }
        proceed.run();
    }

    @PermissionCallback
    private void permCallback(PluginCall call) {
        if (getPermissionState("nearby") == PermissionState.GRANTED
                || getPermissionState("location") == PermissionState.GRANTED) {
            String method = call.getMethodName();
            if ("publish".equals(method))
                publish(call);
            else if ("subscribe".equals(method))
                subscribe(call);
            else
                call.resolve();
        } else {
            call.reject("Permission denied");
        }
    }

    // ==== helpers ====

    private JSObject resultToJS(WifiAwareStateReceiver.AttachResult r) {
        JSObject js = new JSObject();
        js.put("available", r.available);
        if (r.reason != null)
            js.put("reason", r.reason);
        if (r.androidApiLevel != null)
            js.put("androidApiLevel", r.androidApiLevel);
        if (r.instantCommSupported != null)
            js.put("instantCommSupported", r.instantCommSupported);
        return js;
    }

    private JSObject peerToJS(WifiAwareShim.PeerFound p, String serviceName) {
        JSObject js = new JSObject();
        js.put("peerId", p.peerId);
        js.put("serviceName", serviceName);
        if (p.serviceInfoBase64 != null)
            js.put("serviceInfoBase64", p.serviceInfoBase64);
        if (p.distanceMm != null)
            js.put("distanceMm", p.distanceMm);
        return js;
    }

    private JSObject socketToJS(WifiAwareShim.SocketInfo info) {
        JSObject js = new JSObject();
        js.put("role", info.role);
        if (info.localPort != null)
            js.put("localPort", info.localPort);
        if (info.peerIpv6 != null)
            js.put("peerIpv6", info.peerIpv6);
        if (info.peerPort != null)
            js.put("peerPort", info.peerPort);
        return js;
    }

    // ==== Options containers ====

    static class PublishOptions {
        final String serviceName;
        final String serviceInfoBase64;
        final boolean instantMode;
        final boolean rangingEnabled;

        PublishOptions(String s, String info, boolean instant, boolean ranging) {
            this.serviceName = s;
            this.serviceInfoBase64 = info;
            this.instantMode = instant;
            this.rangingEnabled = ranging;
        }

        static PublishOptions fromCall(PluginCall call) {
            String name = call.getString("serviceName");
            if (name == null)
                throw new IllegalArgumentException("serviceName required");
            return new PublishOptions(
                    name,
                    call.getString("serviceInfoBase64"),
                    call.getBoolean("instantMode", false),
                    call.getBoolean("rangingEnabled", false));
        }
    }

    static class SubscribeOptions {
        final String serviceName;
        final boolean instantMode;
        final Integer minDistanceMm;
        final Integer maxDistanceMm;

        SubscribeOptions(String s, boolean instant, Integer min, Integer max) {
            this.serviceName = s;
            this.instantMode = instant;
            this.minDistanceMm = min;
            this.maxDistanceMm = max;
        }

        static SubscribeOptions fromCall(PluginCall call) {
            String name = call.getString("serviceName");
            if (name == null)
                throw new IllegalArgumentException("serviceName required");
            return new SubscribeOptions(
                    name,
                    call.getBoolean("instantMode", false),
                    call.getInt("minDistanceMm"),
                    call.getInt("maxDistanceMm"));
        }
    }
}
